packages/react-menu/src/components/Menu/useMenu.tsx

```diff
@@ -78,19 +78,28 @@ export const useMenu = (props: MenuProps, ref: React.Ref<HTMLElement>, defaultPr
   const [open, setOpen] = useControllableValue(state.open, state.defaultOpen);
   // TODO fix useControllableValue typing
   state.open = open !== undefined ? open : state.open;
+  const { onOpenChange } = state;
   state.setOpen = React.useCallback(
-    (...args) => {
-      setOpen(...args);
+    (e, shouldOpen) => {
+      setOpen(prevOpen => {
+        // More than one event (mouse, focus, keyboard) can request the popup to close
+        // We assume the first event is the correct one
+        if (prevOpen !== shouldOpen) {
+          onOpenChange?.(e, { open: shouldOpen });
+        }
+
+        return shouldOpen;
+      });
     },
-    [setOpen],
+    [setOpen, onOpenChange],
```

Comment:
`onOpenChange` is user's input, correct? In this case we should use `useEventCallback`